Add the Barker-Foran cone to the cone catalog

orlitzky opened the pull request:
This is a handy example of a self-dual (but not symmetric) polyhedral cone given in [Self-dual cones in euclidean spaces](https://www.sciencedirect.com/science/article/pii/0024379576900537?via%3Dihub) by Barker & Foran. I mention it in [Gaddum's test for symmetric cones](https://michael.orlitzky.com/documents/papers/gaddums_test_for_symmetric_cones.pdf), which is freely available.

I've called it the Barker-Foran cone for lack of a better name. The authors define another, related family of cones in the same paper... but the coordinates of their extreme rays aren't rational, so there is no immediate risk of them being added to sage and causing a name clash.

orlitzky committed:
src/doc/en/reference/references/index.rst: add Barker-Foran

Among other things, "Self-Dual Cones in Euclidean Spaces" by Barker &
Foran gives an example of a rational, self-dual convex cone with five
extreme rays in three dimensions. This will soon be included in the
cone catalog.

orlitzky committed:
src/sage/geometry/cone_catalog.py: add barker_foran()

Though it is a name I just made up, the newly minted Barker-Foran cone
is a polyhedral self-dual convex cone with five extreme rays in three
dimensions. Barker & Foran give it as an example in their paper, which
we cite.

It is a nice example because asymmetric self-dual cones are hard to
come by.

orlitzky committed:
src/sage/geometry/cone_catalog.py: disable ruff on one line

The extreme rays of the Barker-Foran cone look much nicer if we
manually align them, but this causes ruff to complain about the one
instance where two spaces appear before a negative sign. We add a
"noqa" comment to disable that check on that one line.

cxzhong commented on src/sage/geometry/cone_catalog.py:
```diff
@@ -161,6 +162,83 @@ def _preprocess_args(ambient_dim, lattice):
+           ( zero,  one,  one),
+           (-one,   zero, one),
+           ( zero, -one,  one),
+           ( one,  -one,  one)]  # noqa: E221
```
Why do you add ignore of E221

orlitzky replied:
I think

```python
ext = [( one,   one,  one),
       ( zero,  one,  one),
       (-one,   zero, one),
       ( zero, -one,  one),
       ( one,  -one,  one)]
```

looks a lot better than

```python
ext = [(one, one, one),
       (zero, one, one),
       (-one, zero, one),
       (zero, -one, one),
       (one, -one, one)]
```

but without the "noqa" the lint (ruff) will fail on mine.

I'll change it if it ruins anyone's day, but in this case I think the lint is just being counterproductive.

cxzhong replied:
Can we directly use 1 and 0?

cxzhong replied:
```
sage: type(ZZ.one())
<class 'sage.rings.integer.Integer'>
sage: type(1)
<class 'sage.rings.integer.Integer'>
```

orlitzky replied:
That only works in a sage shell. The preprocessor turns `1` into `Integer(1)` behind the scenes:

```
sage: preparse("1")
'Integer(1)'
```

We could probably get away with it still, but using `ZZ.zero()` and `ZZ.one()` helps guarantee that we get integer rays.

orlitzky committed:
src/sage/geometry/cone_catalog.py: construct Barker-Foran with check=False

There's no need to check the input for correctness when constructing
the barker_foran() cone, because we know that the input rays all have
three integer components and form a minimal generating set for the
resulting cone. (If you doubt this, construct the cone with
check=True, and look at its generators.)
